### ISSUE TITLE
piVCCU kernel modules unavailable after Raspberry Pi kernel update

### ISSUE THREAD

**Issue by dallmair** [alexreinert/piVCCU#368]
Hi,

Today I ran `sudo apt update && sudo apt dist-upgrade` on my Raspberry Pi OS on Arm64. It upgraded both kernel and matching kernel headers (from `1.20210111.master-1` to `1.20210201-1`) and piVCCU3 (from `3.53.34-51` to `3.55.10-54`). Unfortunately, after this the kernel modules couldn't be loaded anymore:

```
pi@raspberrypi:~ $ sudo pivccu-info
piVCCU version: 3.55.10-54
Kernel modules: Not available
Raw UART dev:   Not available
Rasp.Pi UART:   Assigned to GPIO pins
HMRF Hardware:  FAKE_HMRF
 Board serial:  FKE6956676
 Radio MAC:     0xff7cb9
HMIP Hardware:  HMIP-RFUSB
 SGTIN:         XXX
 Radio MAC:     YYY
State:          STOPPED
```

Journal:
```
Feb 12 21:26:02 raspberrypi systemd[1]: Starting piVCCU...
Feb 12 21:26:02 raspberrypi start_container.sh[3642]: modprobe: FATAL: Module eq3_char_loop not found in directory /lib/modules/5.10.11-v8+
Feb 12 21:26:07 raspberrypi start_container.sh[3642]: /var/lib/piVCCU3/detect_hardware.inc: Zeile 43: /sys/module/fake_hmrf/parameters/firmware_version: Datei oder Verzeichni
Feb 12 21:26:07 raspberrypi start_container.sh[3642]: grep: Schreibfehler: Datenübergabe unterbrochen (broken pipe)
Feb 12 21:26:07 raspberrypi start_container.sh[3642]: cat: /sys/module/fake_hmrf/parameters/firmware_version: Datei oder Verzeichnis nicht gefunden
Feb 12 21:26:11 raspberrypi start_container.sh[3642]: modprobe: FATAL: Module rpi_rf_mod_led not found in directory /lib/modules/5.10.11-v8+
Feb 12 21:26:12 raspberrypi start_container.sh[3642]: kernel.sched_rt_runtime_us = -1
Feb 12 21:26:12 raspberrypi start_container.sh[3642]: lxc-start: lxc: lxccontainer.c: wait_on_daemonized_start: 842 Received container state "ABORTING" instead of "RUNNING"
Feb 12 21:26:12 raspberrypi start_container.sh[3642]: lxc-start: lxc: tools/lxc_start.c: main: 330 The container failed to start
Feb 12 21:26:12 raspberrypi start_container.sh[3642]: lxc-start: lxc: tools/lxc_start.c: main: 333 To get more details, run the container in foreground mode
Feb 12 21:26:12 raspberrypi start_container.sh[3642]: lxc-start: lxc: tools/lxc_start.c: main: 336 Additional information can be obtained by setting the --logfile and --logpr
Feb 12 21:26:12 raspberrypi systemd[1]: Started piVCCU.
Feb 12 21:26:12 raspberrypi systemd[1]: pivccu.service: Main process exited, code=exited, status=1/FAILURE
Feb 12 21:26:12 raspberrypi systemd[1]: pivccu.service: Failed with result 'exit-code'.
```

So I guess the Foundation updated the kernel from 5.4 to 5.10 and with this, the DKMS modules vanished. A simple `sudo apt reinstall pivccu-modules-dkms` did the trick and after `sudo systemctl start pivccu` everything is up and running again :)

However, I'm wondering whether there's a way to automate this. I mean, surely pivccu-modules-dkms plugs into DKMS such that all modules are rebuilt after a kernel/header update, but for whatever reason it didn't work in my case (I'm not too deep into this and didn't really try to identify the root cause). Thus, I'm opening this issue hoping that there's a way to automatically fix such a situation and prevent this from happening again.

**Comment by alexreinert** [comment]
It is already automated. Unfortunately on Raspbian/Raspberry Pi OS, there is only one kernel and one header package installed and not two ore more like it is on x64 Denian or Ubuntu. Because of this, after a kernel update, there will be a mismatch between the running kernel and the kernel headers in the file system until the system is rebooted and the new kernel gets active. If there is an update of the dkms package in that time before the reboot, the kernel modules can not be built for the running kernel because of that mismatch.
There is a startup script which will build the kernel modules after a reboot, if they are not available.

TL;DR; Reboot after each kernel update and it should work.

**Comment by dallmair** [comment]
Wow, many thanks for the quick reply and explanation! Very good to know it's already automated and a simple reboot would have sufficed.

In the meantime I checked my remote location, and it seems that didn't work there, although the system had rebooted automatically :( (nobody realized this as nobody has been in the club house for quite some time due to the current situation)

That one over there is set up with `unattended-upgrades` and `/etc/apt/apt.conf.d/99custom-unattended-upgrades` as follows:
```
Unattended-Upgrade::Origins-Pattern {
        "origin=*";
};
Unattended-Upgrade::Automatic-Reboot "true";
Unattended-Upgrade::Automatic-Reboot-Time "03:30";
```

It installed the updates already on Feb 5:
```
pi@raspberrypi:~ $ cat /var/log/apt/history.log

[...]

Start-Date: 2021-02-05  06:31:20
Commandline: /usr/bin/unattended-upgrade
Upgrade: raspberrypi-kernel-headers:arm64 (1.20210111.master-1, 1.20210201-1)
End-Date: 2021-02-05  06:32:11

[...]

Start-Date: 2021-02-05  06:32:18
Commandline: /usr/bin/unattended-upgrade
Upgrade: pivccu3:arm64 (3.53.34-51, 3.55.10-54)
End-Date: 2021-02-05  06:33:29

Start-Date: 2021-02-05  06:33:33
Commandline: /usr/bin/unattended-upgrade
Upgrade: raspberrypi-kernel:arm64 (1.20210111.master-1, 1.20210201-1)
End-Date: 2021-02-05  06:34:36

[...]
```

And rebooted the following night:

```
pi@raspberrypi:~ $ uptime -s
2021-02-06 03:30:24
```

Per your description, after the automatic reboot the startup script should have run the module build, but piVCCU3 was down with the same symptom and solution. Now I have to admit I had already applied the same reinstall+start sequence before I read your answer, so I guess I unintentionally destroyed part of the system state already :(

Is there anything else I can check on the system regarding this, or shall we just close this issue as a one-off?

**Comment by dallmair** [comment]
Happened again. First kernel and headers were updated:

```
pi@vraspi:~ $ cat /var/log/apt/history.log

Start-Date: 2021-03-07  05:25:16
Commandline: /usr/bin/unattended-upgrade
Upgrade: raspberrypi-bootloader:arm64 (1.20210201-1, 1.20210303-1)
End-Date: 2021-03-07  05:25:24

Start-Date: 2021-03-07  05:25:29
Commandline: /usr/bin/unattended-upgrade
Upgrade: raspberrypi-kernel-headers:arm64 (1.20210201-1, 1.20210303-1)
End-Date: 2021-03-07  05:26:19

Start-Date: 2021-03-07  05:26:25
Commandline: /usr/bin/unattended-upgrade
Upgrade: raspberrypi-kernel:arm64 (1.20210201-1, 1.20210303-1)
End-Date: 2021-03-07  05:27:27
```

Then the system rebooted automatically:

```
pi@vraspi:~ $ uptime -s
2021-03-07 05:28:02
```

But piVCCU didn't come up and the kernel modules have seemingly not been updated:

```
pi@vraspi:~ $ journalctl -u pivccu
-- Logs begin at Sun 2021-03-07 05:27:50 CET, end at Tue 2021-03-09 21:17:01 CET. --
Mär 07 05:30:21 vraspi systemd[1]: Starting piVCCU...
Mär 07 05:30:21 vraspi start_container.sh[19409]: modprobe: FATAL: Module eq3_char_loop not found in directory /lib/modules/5.10.17-v8+
Mär 07 05:30:27 vraspi start_container.sh[19409]: /var/lib/piVCCU3/detect_hardware.inc: Zeile 43: /sys/module/fake_hmrf/parameters/firmware_version: Datei oder Verzeichnis nicht gefunden
Mär 07 05:30:27 vraspi start_container.sh[19409]: cat: /sys/module/fake_hmrf/parameters/firmware_version: Datei oder Verzeichnis nicht gefunden
Mär 07 05:30:36 vraspi start_container.sh[19409]: modprobe: FATAL: Module rpi_rf_mod_led not found in directory /lib/modules/5.10.17-v8+
Mär 07 05:30:38 vraspi start_container.sh[19409]: kernel.sched_rt_runtime_us = -1
Mär 07 05:30:38 vraspi start_container.sh[19409]: lxc-start: lxc: lxccontainer.c: wait_on_daemonized_start: 842 Received container state "ABORTING" instead of "RUNNING"
Mär 07 05:30:38 vraspi start_container.sh[19409]: lxc-start: lxc: tools/lxc_start.c: main: 330 The container failed to start
Mär 07 05:30:38 vraspi start_container.sh[19409]: lxc-start: lxc: tools/lxc_start.c: main: 333 To get more details, run the container in foreground mode
Mär 07 05:30:38 vraspi start_container.sh[19409]: lxc-start: lxc: tools/lxc_start.c: main: 336 Additional information can be obtained by setting the --logfile and --logpriority options
Mär 07 05:30:38 vraspi systemd[1]: Started piVCCU.
Mär 07 05:30:38 vraspi systemd[1]: pivccu.service: Main process exited, code=exited, status=1/FAILURE
Mär 07 05:30:38 vraspi systemd[1]: pivccu.service: Failed with result 'exit-code'.
```

**Comment by dallmair** [comment]
This is actually an issue in DKMS. Here's the reference:

https://github.com/dell/dkms/issues/83

Thus, I'm closing the this one now.

**Comment by dallmair** [comment]
Hmm, this happened again today and I had some time to investigate, and I think piVCCU is in a good position to carry a workaround. So I'm reopening this issue for discussion.

In the last months, the RasPi Foundation pushed a bunch of kernel updates, and installation of most of them worked fine for me. Digging through the logs, it seems to me now that the installation sequence `kernel -> headers -> reboot` works fine, but the sequence `headers -> kernel -> reboot` causes a broken system in the sense that even after the reboot the kernel modules cannot be loaded and therefore piVCCU doesn't start.

The modules have been built and installed just fine:

```
pi@vraspi:~ $ dkms status
pivccu, 1.0.63, 5.10.60-v8+, aarch64: installed
```

But the modules cannot be loaded, and `ensure_modules.sh` detects and tries to fix the situation (if the `modinfo` would have worked, then `ensure_modules.sh` wouldn't even try to call `dkms install`, so `modinfo` must have failed):

```
pi@vraspi:~ $ journalctl -u pivccu-dkms
-- Logs begin at Sun 2021-09-05 05:25:52 CEST, end at Mon 2021-09-06 01:35:34 CEST. --
Sep 05 05:25:55 vraspi systemd[1]: Starting piVCCU DKMS Modules...
Sep 05 05:25:56 vraspi ensure_modules.sh[322]: Module pivccu/1.0.63 already installed on kernel 5.10.60-v8+/aarch64
Sep 05 05:25:56 vraspi systemd[1]: pivccu-dkms.service: Succeeded.
Sep 05 05:25:56 vraspi systemd[1]: Started piVCCU DKMS Modules.
```

After a lot of trial & error, I discovered that simply running `depmod` again (but obviously running with the new kernel) fixes the module load problem:

```
pi@vraspi:~ $ sudo depmod -a 5.10.60-v8+
pi@vraspi:~ $ sudo modprobe generic_raw_uart
pi@vraspi:~ $ sudo systemctl start pivccu
pi@vraspi:~ $
```

Would it be possible to include something similar to [`do_depmod`](https://github.com/dell/dkms/blob/v2.8.6/dkms#L300-L308) in [`ensure_modules.sh`](https://github.com/alexreinert/piVCCU/blob/master/pivccu/dkms/ensure_modules.sh#L34)? That change should be low risk, I think, and has the potential to make updates more reliable.

Thanks!

**Comment by dallmair** [comment]
As there's no progress here, let me simply post my local workaround. Maybe it helps somebody until there is an official fix from DKMS or a workaround included in piVCCU. Here's how I plugged the proposed changes into a system without modification of existing files, such that piVCCU updates can be installed seamlessly and in fully automated fashion on a Raspberry Pi. Worked flawlessly for the last few months :)

/etc/systemd/system/pivccu-depmod.service
```
[Unit]
Description=piVCCU depmod caller
Before=pivccu.service
Before=debmatic.service
After=pivccu-dkms.service
Wants=pivccu-dkms.service

[Service]
ExecStart=/usr/local/bin/pivccu-depmod
Type=oneshot
TimeoutSec=600s

[Install]
WantedBy=multi-user.target
RequiredBy=pivccu.service
WantedBy=debmatic.service
```

/usr/local/bin/pivccu-depmod
```
#!/bin/bash

modinfo generic_raw_uart &> /dev/null

if [ ! $? -eq 0 ]; then
  PKG_VER=`dpkg -s pivccu-modules-dkms | grep '^Version: ' | cut -d' ' -f2`

  if [[ "$(dkms status -m pivccu -v $PKG_VER -k `uname -r`)" =~ "installed" ]]; then
    echo Running depmod
    if [ "$(uname -s)" == "Linux" ] ; then
      if [[ -f /boot/System.map-`uname -r` ]]; then
        depmod -a "$(uname -r)" -F "/boot/System.map-$(uname -r)"
      else
        depmod -a "$(uname -r)"
      fi
    fi
  fi
fi
```

To explain the approach in simple words: I extracted the workaround in the linked PR to a dedicated systemd service. This means all piVCCU files are untouched on the system, and therefore updates can be installed seamlessly without the package manager complaining about locally modified files. And by conditionally running `depmod` before `pivccu.service`, piVCCU starts successfully regardless of whether the kernel itself or the kernel headers were installed at the last update.

**Comment by alexreinert** [comment]
Fixed in commit 0eba85aand released in pivccu-modules-dkms version 1.0.68 in the testing APT repository

**Comment by dallmair** [comment]
Yours is an even easier and (I'm sure) more stable fix, many thanks for that!
